### ISSUE TITLE
How to run health checks on Signalbot?

### ISSUE THREAD

**Issue by ThorpeJosh** [filipre/signalbot#26]
Hi, Thanks for a great library.

I built a signalbot that replies to group messages and deployed it in a docker container along side signal-api.

The issue I have encountered is randomly a couple of times the signalbot has locked up, frozen and unresponsive. A simple restart of the container fixes the issue, however it keeps happening.

I created a healthcheck with the following that simply writes the current UTC epoch time to `/tmp/healthcheck` then docker checks every 60s to make sure the time in that file is not more that 60s old else it presumes the signalbot has frozen and restarts the container.
```python
    # Docker healthcheck
    def healthcheck():
        # Write the current UTC epoch time to file
        with open("/tmp/healthcheck", mode="w", encoding="utf-8") as file:
            file.write(str(int(time.time())))

    bot.scheduler.add_job(healthcheck, "interval", seconds=60)
```

This fixed the issue for a few months, however just today I encountered an issue where the bot stopped working again, but the healthcheck still worked and continued to write to the file every 60s. I could send/receive messages using curl directly to the signal-api so no issues there, and as soon as I manually restarted the signal-bot container, it worked correctly again.

Do you have any ideas on how to more reliably detect the health of the signalbot?

**Comment by filipre** [comment]
>The issue I have encountered is randomly a couple of times the signalbot has locked up, frozen and unresponsive. A simple restart of the container fixes the issue, however it keeps happening.

That's interesting. I noticed this so far only once since I have it running. Do you have any logs to share so we could identify the root cause?

>but the healthcheck still worked and continued to write to the file every 60s.

This makes sense, because the scheduler is a third party package in reality running on the same event loop. I suppose the `producer` stopped working but not sure why. Maybe the socket connection to the signal-api has been closed somehow?

I used a very similar method to detect downtimes where I called a uptime-kuma health check URL
```python
self.bot.scheduler.add_job(
    self._call_push_url,
    "interval",
    seconds=self.call_interval,
    id="renebot_util_healthcheck",
    replace_existing=True,
)
```

I could imagine that you can somehow write a health check that simulates sending a message and checks if it has been reacted on. Can you experiment with this command? In theory, the `listen` flag should enable sending messages to itself. But I remember I had some issues with it, too.
```python
await self.bot.send(receiver, msg, listen=True)
```

**Comment by ThorpeJosh** [comment]
So looking back through the logs I now remember that each time signalbot stopped working there generally were tracebacks however the process never ended so the container stayed alive and would never get restarted by docker.

I just found the traceback for the latest crash and as you predicted I think it was a failed connection to the signal-api.
```
2023-06-30T22:11:40.194152707Z 2023-07-01 06:11:40,193 - root - INFO - [Bot] Producer #1 started
2023-06-30T22:11:40.860016816Z 2023-07-01 06:11:40,364 - root - INFO - [Bot] Consumer #1 started
2023-06-30T22:11:40.860113904Z 2023-07-01 06:11:40,364 - root - INFO - [Bot] Consumer #2 started
2023-06-30T22:11:40.860130351Z 2023-07-01 06:11:40,364 - root - INFO - [Bot] Consumer #3 started
2023-06-30T22:11:40.882718274Z 2023-07-01 06:11:40,856 - asyncio - ERROR - Task exception was never retrieved
2023-06-30T22:11:40.882779711Z future: <Task finished name='Task-1' coro=<SignalBot._produce_consume_messages() done, defined at /usr/local/lib/python3.11/site-packages/signalbot/bot.py:219> exception=SignalBotError("Cannot receive messages: [Errno 111] Connect call failed ('172.100.0.3', 8080)")>
2023-06-30T22:11:40.882797850Z Traceback (most recent call last):
2023-06-30T22:11:40.882810576Z   File "/usr/local/lib/python3.11/site-packages/signalbot/api.py", line 20, in receive
2023-06-30T22:11:40.882823349Z     async with self.connection as websocket:
2023-06-30T22:11:40.882836224Z   File "/usr/local/lib/python3.11/site-packages/websockets/legacy/client.py", line 642, in __aenter__
2023-06-30T22:11:40.882848982Z     return await self
2023-06-30T22:11:40.882861165Z            ^^^^^^^^^^
2023-06-30T22:11:40.882873417Z   File "/usr/local/lib/python3.11/site-packages/websockets/legacy/client.py", line 659, in __await_impl_timeout__
2023-06-30T22:11:40.882900349Z     return await asyncio.wait_for(self.__await_impl__(), self.open_timeout)
2023-06-30T22:11:40.882914130Z            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-06-30T22:11:40.882926693Z   File "/usr/local/lib/python3.11/asyncio/tasks.py", line 479, in wait_for
2023-06-30T22:11:40.882939265Z     return fut.result()
2023-06-30T22:11:40.882951622Z            ^^^^^^^^^^^^
2023-06-30T22:11:40.882963875Z   File "/usr/local/lib/python3.11/site-packages/websockets/legacy/client.py", line 663, in __await_impl__
2023-06-30T22:11:40.882976507Z     _transport, _protocol = await self._create_connection()
2023-06-30T22:11:40.882988747Z                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-06-30T22:11:40.883000978Z   File "/usr/local/lib/python3.11/asyncio/base_events.py", line 1085, in create_connection
2023-06-30T22:11:40.883013435Z     raise exceptions[0]
2023-06-30T22:11:40.883025497Z   File "/usr/local/lib/python3.11/asyncio/base_events.py", line 1069, in create_connection
2023-06-30T22:11:40.883038039Z     sock = await self._connect_sock(
2023-06-30T22:11:40.883050252Z            ^^^^^^^^^^^^^^^^^^^^^^^^^
2023-06-30T22:11:40.883062510Z   File "/usr/local/lib/python3.11/asyncio/base_events.py", line 973, in _connect_sock
2023-06-30T22:11:40.883075062Z     await self.sock_connect(sock, address)
2023-06-30T22:11:40.883087287Z   File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 634, in sock_connect
2023-06-30T22:11:40.883099813Z     return await fut
2023-06-30T22:11:40.883111928Z            ^^^^^^^^^
2023-06-30T22:11:40.883124052Z   File "/usr/local/lib/python3.11/asyncio/selector_events.py", line 674, in _sock_connect_cb
2023-06-30T22:11:40.883136683Z     raise OSError(err, f'Connect call failed {address}')
2023-06-30T22:11:40.883149069Z ConnectionRefusedError: [Errno 111] Connect call failed ('172.100.0.3', 8080)
2023-06-30T22:11:40.883161498Z 
2023-06-30T22:11:40.883175683Z During handling of the above exception, another exception occurred:
2023-06-30T22:11:40.883188834Z 
2023-06-30T22:11:40.883200889Z Traceback (most recent call last):
2023-06-30T22:11:40.883213004Z   File "/usr/local/lib/python3.11/site-packages/signalbot/bot.py", line 234, in _produce
2023-06-30T22:11:40.883225607Z     async for raw_message in self._signal.receive():
2023-06-30T22:11:40.883237990Z   File "/usr/local/lib/python3.11/site-packages/signalbot/api.py", line 25, in receive
2023-06-30T22:11:40.883250843Z     raise ReceiveMessagesError(e)
2023-06-30T22:11:40.883263087Z signalbot.api.ReceiveMessagesError: [Errno 111] Connect call failed ('172.100.0.3', 8080)
2023-06-30T22:11:40.883275477Z 
2023-06-30T22:11:40.883287492Z During handling of the above exception, another exception occurred:
2023-06-30T22:11:40.883304336Z 
2023-06-30T22:11:40.883332720Z Traceback (most recent call last):
2023-06-30T22:11:40.883345865Z   File "/usr/local/lib/python3.11/site-packages/signalbot/bot.py", line 226, in _produce_consume_messages
2023-06-30T22:11:40.883359013Z     await asyncio.gather(*producers)
2023-06-30T22:11:40.883371338Z   File "/usr/local/lib/python3.11/site-packages/signalbot/bot.py", line 249, in _produce
2023-06-30T22:11:40.883384068Z     raise SignalBotError(f"Cannot receive messages: {e}")
2023-06-30T22:11:40.883396753Z signalbot.bot.SignalBotError: Cannot receive messages: [Errno 111] Connect call failed ('172.100.0.3', 8080)
2023-06-30T22:12:40.240321527Z 2023-07-01 06:12:40,239 - apscheduler.scheduler - DEBUG - Looking for jobs to run
2023-06-30T22:12:40.241150136Z 2023-07-01 06:12:40,240 - apscheduler.scheduler - DEBUG - Next wakeup is due at 2023-07-01 06:13:40.179170+08:00 (in 59.938737 seconds)
2023-06-30T22:12:40.241853795Z 2023-07-01 06:12:40,241 - apscheduler.executors.default - INFO - Running job "main.<locals>.healthcheck (trigger: interval[0:01:00], next run at: 2023-07-01 06:13:40 +08)" (scheduled at 2023-07-01 06:12:40.179170+08:00)
2023-06-30T22:12:40.244146690Z 2023-07-01 06:12:40,243 - apscheduler.executors.default - INFO - Job "main.<locals>.healthcheck (trigger: interval[0:01:00], next run at: 2023-07-01 06:13:40 +08)" executed successfully
2023-06-30T22:13:40.243746575Z 2023-07-01 06:13:40,239 - apscheduler.scheduler - DEBUG - Looking for jobs to run
2023-06-30T22:13:40.243819628Z 2023-07-01 06:13:40,240 - apscheduler.scheduler - DEBUG - Next wakeup is due at 2023-07-01 06:14:40.179170+08:00 (in 59.939123 seconds)
2023-06-30T22:13:40.243834868Z 2023-07-01 06:13:40,240 - apscheduler.executors.default - INFO - Running job "main.<locals>.healthcheck (trigger: interval[0:01:00], next run at: 2023-07-01 06:14:40 +08)" (scheduled at 2023-07-01 06:13:40.179170+08:00)
2023-06-30T22:13:40.243850006Z 2023-07-01 06:13:40,241 - apscheduler.executors.default - INFO - Job "main.<locals>.healthcheck (trigger: interval[0:01:00], next run at: 2023-07-01 06:14:40 +08)" executed successfully
2023-06-30T22:14:40.224027643Z 2023-07-01 06:14:40,223 - apscheduler.scheduler - DEBUG - Looking for jobs to run
2023-06-30T22:14:40.225283795Z 2023-07-01 06:14:40,224 - apscheduler.scheduler - DEBUG - Next wakeup is due at 2023-07-01 06:15:40.179170+08:00 (in 59.955313 seconds)
2023-06-30T22:14:40.225329931Z 2023-07-01 06:14:40,224 - apscheduler.executors.default - INFO - Running job "main.<locals>.healthcheck (trigger: interval[0:01:00], next run at: 2023-07-01 06:15:40 +08)" (scheduled at 2023-07-01 06:14:40.179170+08:00)
2023-06-30T22:14:40.228966225Z 2023-07-01 06:14:40,228 - apscheduler.executors.default - INFO - Job "main.<locals>.healthcheck (trigger: interval[0:01:00], next run at: 2023-07-01 06:15:40 +08)" executed successfully
```
I will remove my healthcheck since it clearly isn't working; however, in the past when a traceback occurs the logs would end there but the python process wouldn't exit.

Do you know why the signalbot process stays alive when an exception occurs?

**Comment by filipre** [comment]
Thank you for the logs. Those are very useful!

>Do you know why the signalbot process stays alive when an exception occurs?

I suspect it's because the exception is thrown in a different thread. I'm not sure how I can restart the task but I feel like there must be an easy way to do so. I don't find it unreasonable that sometimes a connection drops. I can debug this once I have more time. If you want to simulate network failures, you could try out https://requestly.io/ . I think that way we can reproduce the bug

**Comment by ThorpeJosh** [comment]
I have the following in my docker compose signalbot service:
```
depends_on:
  signal-api:
    condition: service_healthy
    restart: true
```
It ensures that the signal-api service is up and running before my signalbot app launches.

If I comment out the above code, deploy the signalbot and then restart/stop the signal-api (simulating a websocket connection issue) then it immediately reproduces the error in the same manner as below.
```
2023-07-09T09:00:47.648997302Z 2023-07-09 17:00:47,648 - root - INFO - Starting signal bot
2023-07-09T09:00:47.649399760Z 2023-07-09 17:00:47,649 - asyncio - DEBUG - Using selector: EpollSelector
2023-07-09T09:00:47.650525239Z 2023-07-09 17:00:47,650 - root - WARNING - [Bot] Could not initialize Redis. In-memory storage will be used. Restarting will delete the storage!
2023-07-09T09:00:49.020886941Z 2023-07-09 17:00:49,019 - apscheduler.scheduler - INFO - Scheduler started
2023-07-09T09:00:49.024338438Z 2023-07-09 17:00:49,020 - apscheduler.scheduler - DEBUG - Looking for jobs to run
2023-07-09T09:00:49.024691892Z 2023-07-09 17:00:49,024 - apscheduler.scheduler - DEBUG - No jobs; waiting until a job is added
2023-07-09T09:00:49.025231403Z 2023-07-09 17:00:49,024 - root - INFO - [Bot] Producer #1 started
2023-07-09T09:00:49.079504186Z 2023-07-09 17:00:49,078 - root - INFO - [Bot] Consumer #1 started
2023-07-09T09:00:49.079667031Z 2023-07-09 17:00:49,079 - root - INFO - [Bot] Consumer #2 started
2023-07-09T09:00:49.080266160Z 2023-07-09 17:00:49,079 - root - INFO - [Bot] Consumer #3 started
2023-07-09T09:00:49.094876987Z 2023-07-09 17:00:49,094 - websockets.client - DEBUG - = connection is CONNECTING
2023-07-09T09:00:49.096391171Z 2023-07-09 17:00:49,095 - websockets.client - DEBUG - > GET /v1/receive/+6588888888 HTTP/1.1
2023-07-09T09:00:49.096497573Z 2023-07-09 17:00:49,095 - websockets.client - DEBUG - > Host: signal-api:8080
2023-07-09T09:00:49.097023978Z 2023-07-09 17:00:49,096 - websockets.client - DEBUG - > Upgrade: websocket
2023-07-09T09:00:49.097074442Z 2023-07-09 17:00:49,096 - websockets.client - DEBUG - > Connection: Upgrade
2023-07-09T09:00:49.097094420Z 2023-07-09 17:00:49,096 - websockets.client - DEBUG - > Sec-WebSocket-Key: ABCD/H/abcdabcdabcdabcd==
2023-07-09T09:00:49.097113431Z 2023-07-09 17:00:49,096 - websockets.client - DEBUG - > Sec-WebSocket-Version: 13
2023-07-09T09:00:49.098016641Z 2023-07-09 17:00:49,096 - websockets.client - DEBUG - > Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits
2023-07-09T09:00:49.098126327Z 2023-07-09 17:00:49,096 - websockets.client - DEBUG - > User-Agent: Python/3.11 websockets/10.4
2023-07-09T09:00:49.098881738Z 2023-07-09 17:00:49,098 - websockets.client - DEBUG - < HTTP/1.1 101 Switching Protocols
2023-07-09T09:00:49.098916671Z 2023-07-09 17:00:49,098 - websockets.client - DEBUG - < Upgrade: websocket
2023-07-09T09:00:49.099265143Z 2023-07-09 17:00:49,098 - websockets.client - DEBUG - < Connection: Upgrade
2023-07-09T09:00:49.099290712Z 2023-07-09 17:00:49,098 - websockets.client - DEBUG - < Sec-WebSocket-Accept: sk0UNxNsXsyC+MREpX477p3jn9M=
2023-07-09T09:00:49.099710851Z 2023-07-09 17:00:49,099 - websockets.client - DEBUG - = connection is OPEN
2023-07-09T09:01:27.022854919Z 2023-07-09 17:01:27,022 - websockets.client - DEBUG - ! failing connection with code 1006
2023-07-09T09:01:27.023224917Z 2023-07-09 17:01:27,022 - websockets.client - DEBUG - = connection is CLOSED
2023-07-09T09:01:27.038693227Z 2023-07-09 17:01:27,024 - asyncio - ERROR - Task exception was never retrieved
2023-07-09T09:01:27.038768250Z future: <Task finished name='Task-1' coro=<SignalBot._produce_consume_messages() done, defined at /usr/local/lib/python3.11/site-packages/signalbot/bot.py:219> exception=SignalBotError('Cannot receive messages: no close frame received or sent')>
2023-07-09T09:01:27.038786052Z Traceback (most recent call last):
2023-07-09T09:01:27.038800000Z   File "/usr/local/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 968, in transfer_data
2023-07-09T09:01:27.038814471Z     message = await self.read_message()
2023-07-09T09:01:27.038828200Z               ^^^^^^^^^^^^^^^^^^^^^^^^^
2023-07-09T09:01:27.039037619Z   File "/usr/local/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 1038, in read_message
2023-07-09T09:01:27.039117275Z     frame = await self.read_data_frame(max_size=self.max_size)
2023-07-09T09:01:27.039132718Z             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-07-09T09:01:27.039146438Z   File "/usr/local/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 1113, in read_data_frame
2023-07-09T09:01:27.039160651Z     frame = await self.read_frame(max_size)
2023-07-09T09:01:27.039322043Z             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2023-07-09T09:01:27.039340495Z   File "/usr/local/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 1170, in read_frame
2023-07-09T09:01:27.039355119Z     frame = await Frame.read(
2023-07-09T09:01:27.039368455Z             ^^^^^^^^^^^^^^^^^
2023-07-09T09:01:27.039399902Z   File "/usr/local/lib/python3.11/site-packages/websockets/legacy/framing.py", line 69, in read
2023-07-09T09:01:27.039414669Z     data = await reader(2)
2023-07-09T09:01:27.039427946Z            ^^^^^^^^^^^^^^^
2023-07-09T09:01:27.039531046Z   File "/usr/local/lib/python3.11/asyncio/streams.py", line 727, in readexactly
2023-07-09T09:01:27.040002251Z     raise exceptions.IncompleteReadError(incomplete, n)
2023-07-09T09:01:27.040019960Z asyncio.exceptions.IncompleteReadError: 0 bytes read on a total of 2 expected bytes
2023-07-09T09:01:27.040033947Z 
2023-07-09T09:01:27.040061426Z The above exception was the direct cause of the following exception:
2023-07-09T09:01:27.040079013Z 
2023-07-09T09:01:27.040092575Z Traceback (most recent call last):
2023-07-09T09:01:27.040105826Z   File "/usr/local/lib/python3.11/site-packages/signalbot/api.py", line 21, in receive
2023-07-09T09:01:27.040119663Z     async for raw_message in websocket:
2023-07-09T09:01:27.040132873Z   File "/usr/local/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 497, in __aiter__
2023-07-09T09:01:27.040146728Z     yield await self.recv()
2023-07-09T09:01:27.040159844Z           ^^^^^^^^^^^^^^^^^
2023-07-09T09:01:27.040173049Z   File "/usr/local/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 568, in recv
2023-07-09T09:01:27.040186688Z     await self.ensure_open()
2023-07-09T09:01:27.040206727Z   File "/usr/local/lib/python3.11/site-packages/websockets/legacy/protocol.py", line 944, in ensure_open
2023-07-09T09:01:27.040308170Z     raise self.connection_closed_exc()
2023-07-09T09:01:27.041369583Z websockets.exceptions.ConnectionClosedError: no close frame received or sent
2023-07-09T09:01:27.041386083Z 
2023-07-09T09:01:27.041399655Z During handling of the above exception, another exception occurred:
2023-07-09T09:01:27.041429830Z 
2023-07-09T09:01:27.041496262Z Traceback (most recent call last):
2023-07-09T09:01:27.041511549Z   File "/usr/local/lib/python3.11/site-packages/signalbot/bot.py", line 234, in _produce
2023-07-09T09:01:27.041525652Z     async for raw_message in self._signal.receive():
2023-07-09T09:01:27.041539057Z   File "/usr/local/lib/python3.11/site-packages/signalbot/api.py", line 25, in receive
2023-07-09T09:01:27.041567663Z     raise ReceiveMessagesError(e)
2023-07-09T09:01:27.041582563Z signalbot.api.ReceiveMessagesError: no close frame received or sent
2023-07-09T09:01:27.041595961Z 
2023-07-09T09:01:27.041609162Z During handling of the above exception, another exception occurred:
2023-07-09T09:01:27.041623075Z 
2023-07-09T09:01:27.041636111Z Traceback (most recent call last):
2023-07-09T09:01:27.041649377Z   File "/usr/local/lib/python3.11/site-packages/signalbot/bot.py", line 226, in _produce_consume_messages
2023-07-09T09:01:27.041662949Z     await asyncio.gather(*producers)
2023-07-09T09:01:27.041676166Z   File "/usr/local/lib/python3.11/site-packages/signalbot/bot.py", line 249, in _produce
2023-07-09T09:01:27.041689674Z     raise SignalBotError(f"Cannot receive messages: {e}")
2023-07-09T09:01:27.041712782Z signalbot.bot.SignalBotError: Cannot receive messages: no close frame received or sent
```

The big issue is that the signalbot never exits, it just hangs there indefinitely. 
You can see after the error is thrown the python process is still sitting idle.
```bash
host$ docker compose exec -it signal-bot sh
signalbot$ ps -a
PID   USER     TIME  COMMAND
1 root      0:00 tini -- /bot.py
7 root      0:01 {bot.py} /usr/local/bin/python /bot.py
15 root      0:00 sh
22 root      0:00 ps -a
```

I tried looking but cannot see what is catching the `signalbot.bot.SignalBotError` exception and stopping the python interpreter from exiting.

**Comment by ThorpeJosh** [comment]
Hey @filipre,

I've done some reading and asynchronous event loops suffer from not handling exceptions that occur in worker threads. I played around and managed to get the event loop to stop (and python to exit) when the SignalBotError is raised via two different methods.

1. Change `run_forever()` to `run_until_complete()`  

Unchanged code:
```python
# Run event loop
self._event_loop.run_forever()
```
Modified:
```python
# Run event loop
self._event_loop.run_until_complete(self._produce_consume_messages())
```
I'm not sure what implications this method has for any scheduled tasks since it now only runs a single task in the loop.


2. Stop the event loop explicitly when a `SignalBotError` is raised

Unchanged code:
```python
async def _produce_consume_messages(self, producers=1, consumers=3) -> None:
    producers = [
      asyncio.create_task(self._produce(n)) for n in range(1, producers + 1)
    ]
    consumers = [
      asyncio.create_task(self._consume(n)) for n in range(1, consumers + 1)
    ]
    await asyncio.gather(*producers)
    await self._q.join()
    for c in consumers:
      c.cancel()
```
Modified:
```python
async def _produce_consume_messages(self, producers=1, consumers=3) -> None:
  try:
    producers = [
      asyncio.create_task(self._produce(n)) for n in range(1, producers + 1)
    ]
    consumers = [
      asyncio.create_task(self._consume(n)) for n in range(1, consumers + 1)
    ]
    await asyncio.gather(*producers)
    await self._q.join()
    for c in consumers:
      c.cancel()
  except SignalBotError:
    self._event_loop.stop()
```
This method stops the loop for any occurance of `SignalBotError`, which is probably not desirable when looking at the types of situations some `SignalBotError`s are being raised.

What do you think of these approaches? 
I'm happy to raise a PR for either method, or try make it configurable. Maybe have an optional parameter for `bot.start(kill_on_exception=false)` or something.

**Comment by filipre** [comment]
Hey, I followed a slightly different approach. Instead of killing the event loop completely, I just restart the producer when the connection to the signal-api stops or when there is another unexpected exception. Could you review https://github.com/filipre/signalbot/pull/27 and let me know what you think? My deploy script does not work but I can provide some test package for you later.

**Comment by filipre** [comment]
You can test the package using this command in your python environment
```
pip uninstall signalbot
pip install -i https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple signalbot==0.7.0a1689425715
```

**Comment by ThorpeJosh** [comment]
Thanks for putting that together. 
I tested using that package and it successfully recovered from websocket disconnects.

I left a comment on the PR otherwise everything looks good.

**Comment by filipre** [comment]
Thank you for the comment. I released the new version `0.7.0`. Let me know if you encounter another problem or if this approach does not work for some reason

**Comment by ThorpeJosh** [comment]
Awesome. Thanks @filipre
